// python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. functions.proto
// https://github.com/xainag/numproto
syntax = "proto3";
import "google/protobuf/struct.proto";

message Model {
    bytes model = 1;
    bytes xtheta = 2;
    string id = 3;
}

message Empty {
    int32 value = 1;
}

message Reply {
  string str_reply = 1;
  double numeric_reply = 2;
}

message Pair {
   string key = 1;
   string value = 2;
}

message config {
    map<string, google.protobuf.Value> params = 1;
}

message Dictionary {
   repeated Pair pairs = 1;
}

message InitialParams {
    double alpha = 1;
    double index  = 2;
    double isDC = 3;
    int32 dc_index = 4;
    int32 device_index = 5;
    int32 coordinate_per_dc = 6;
    int32 datapoints_per_device = 7;
    int32 randomseed = 8;
    double lambduh = 9;
    bool decreasing_step = 10;
    bytes model = 11;
}

message Note {
    string name = 1;
    string message = 2;
}

message Result {
    bytes result = 1;
}

message TrainConfig{
    Model model = 1;
    int32 q = 2;
    double lambduh = 3;
}

message Data{
    bytes x=1;
    bytes y=2;
}

service FederatedApp{
    rpc InitializeParams(InitialParams) returns (Reply) {}
    rpc GenerateData(Empty) returns (Empty) {}
    rpc InitializeData(Data) returns (Reply) {}
    rpc Train(TrainConfig) returns (Model) {}
    rpc Test(Empty) returns (Note) {}
    rpc SendModel(Empty) returns (Model) {}
    rpc UpdateLocalModels(Model) returns (Model) {}
    //rpc ChatStream (Empty) returns (stream Note)
}